frontend/src/firebase/actions.ts

```diff
@@ -373,6 +374,74 @@ export const removeSubTask = (
  * --------------------------------------------------------------------------------
  */
 
+/**
+* Join a group.
+* @param groupID  Document ID of the group's Firestore document. The user calling this function must
+*                 be a member of this group.
+* @param inviteID Document ID of the invitation's Firestore document. The user calling this invitee
+*                 of this invitation.
+*/
+
+export const joinGroup = async (
+  groupId: string,
+  inviteID: string,
+): Promise<void> => {
+  const groupDoc = await database.groupsCollection().doc(groupId);
+  const { groups } = store.getState();
+  const { pendingInvites } = store.getState();
+  const invite = pendingInvites.get(inviteID);
+  // Check if user has the invitation and invitation is for this group
+  if (invite === undefined || invite.group !== groupId) {
+    return;
+  }
+  const members = groups.get(groupId)?.members;
+  const { email } = getAppUser();
+  // Check if user is already in the group
+  if (members === undefined || members.includes(email)) {
+    return;
+  }
+  groupDoc.update({ members: [...members, email] });
+};
```

Comment:
I don't think we're going to be able to directly query Firestore for this operation. There's no real way for us to write the security rules that will allow us to pass in the invite ID to check against without writing it to the document, which we don't want. We might need to use a Cloud Function for this. We can talk about it at the meeting today.

Reply:
We might still be able to do this, but rely on a weaker security model, with the assumption that groupId is unguessable. Then invitations' only purpose is to tell you a specific groupId.

Reply:
Here is an idea that might work:
Store invitation and group in the same document with format like:
```js
{
  invitations: {
    // invited         -> invitee name
    'foo@bar.com': 'Beep Boop'
  },
  members: ['1@a.com', '2@b.com']
}
```

If 'foo@bar.com' accepts the invitation, it will now become:

```js
{
  invitations: { },
  members: ['1@a.com', '2@b.com', 'foo@bar.com']
}
```

The security rule can then do some map and set diff to validate this.
https://firebase.google.com/docs/reference/rules/rules.Map
https://firebase.google.com/docs/reference/rules/rules.Set
https://firebase.google.com/docs/reference/rules/rules.MapDiff